### PR TITLE
Error overlays

### PR DESCRIPTION
**Pull request by deepfire**

### CODE REVIEW

**Comment by deepfire** [comment]
Origin PR: https://github.com/haskell/haskell-mode/pull/766

File: haskell-load.el [hunk 1]
```diff
@@ -265,10 +266,132 @@ actual Emacs buffer of the module being loaded."
...
+	(overlay-put ovl 'haskell-hole hole)
+	(cl-destructuring-bind (face fringe) (cl-case type
+					       (warning (list 'haskell-warning-face haskell-check-warning-fringe))
+					       (hole    (list 'haskell-hole-face    haskell-check-hole-fringe))
+					       (error   (list 'haskell-error-face   haskell-check-error-fringe)))
+	  (overlay-put ovl 'before-string fringe)
+	  (overlay-put ovl 'face face))))))
+
 (defun haskell-process-errors-warnings (module-buffer session process buffer &optional return-only)
   "Trigger handling type errors or warnings. Either prints the
```
Comment (geraldus) [inline haskell-load.el]: Can you please add second whitespace after `.` (emacs lint produces a warnings about this)?

**Comment by geraldus** [comment]
very nice!

**Comment by deepfire** [comment]
@geraldus, pushed, but, if I'm not wrong the lack of a double-space in that docstring wasn't introduced by me.

**Comment by gracjan** [comment]
@deepfire: Good stuff. There are some things to get in shape, overall this is pretty cool thing!

**Comment by deepfire** [comment]
@gracjan, I'm thinking.. is it time to separate overlay functions into `haskell-overlay.el` ?

**Comment by deepfire** [comment]
@gracjan, after a litany of curses I have made `ghc-mod` to work again -- without error-overlays, that is.
I didn't manage to get it working properly, which is a perennial issue -- and indeed, one of my chief motivators to jump ship -- all I got to work was a misplaced error overlay at the top of the file.

And I can happily report that:
1. `error-overlays` does not seem to make this user experience any worser
2. `M-n`/`M-p` doesn't seem to conflict with any `ghc-mod`'s bindings -- and a glance at their source suggests that this is how it should be, indeed

I'm not sure if this constitutes sufficient testing, though.

**Comment by gracjan** [comment]
Thanks.

**Comment by geraldus** [comment]
Cool, thanks!  Let's try this out!

**Comment by deepfire** [comment]
@geraldus, it's somewhat incomplete without the other pull requests, I'm afraid..

**Comment by deepfire** [comment]
@geraldus, at a minimum you would want:
```
(pushnew "-Wall" haskell-process-args-ghci)
```

File: haskell.el [hunk 2]
```diff
@@ -49,6 +49,8 @@
     (define-key map (kbd "C-c C-x") 'haskell-process-cabal)
     (define-key map [?\C-c ?\C-b] 'haskell-interactive-switch)
     (define-key map [?\C-c ?\C-z] 'haskell-interactive-switch)
+    (define-key map (kbd "M-n") 'haskell-goto-next-error)
+    (define-key map (kbd "M-p") 'haskell-goto-prev-error)
```
Comment (DanielG) [inline haskell.el]: Aren't these going to conflict with ghc-mod?

Comment (deepfire) [inline haskell.el]: @DanielG, do you mean with:
```
(defvar ghc-previous-key    "\ep")
(defvar ghc-next-key        "\en")
```
I must say I overlooked this, then..

But then, there's the more pressing question of having two sets of overlays on at the same time, which is still unresolved.

**Pull request merged**